Shell/main.cpp

```diff
@@ -153,6 +153,200 @@ static int sh_umask(int argc, char** argv)
+        int rc = chdir(real_path);
+        if (rc < 0) {
+            printf("chdir(%s) failed: %s\n", real_path, strerror(errno));
+            return 1;
+        }
+
+        g.cwd = canonical_path.string();
+    }
+
+    return 0;
+}
+
+static int sh_pushd(int argc, char** argv)
+{
+    char pathbuf[PATH_MAX];
+    bool should_switch = true;
+
+    // From the BASH reference manual: https://www.gnu.org/software/bash/manual/html_node/Directory-Stack-Builtins.html
+    // With no arguments, pushd exchanges the top two directories and makes the new top the current directory.
+    if (argc == 1) {
+        if (g.directory_stack.size() < 2) {
+            printf("pushd: no other directory\n");
+            return 1;
+        }
+
+        String dir1 = g.directory_stack.take_first();
+        String dir2 = g.directory_stack.take_first();
+        g.directory_stack.insert(0, dir2);
+        g.directory_stack.insert(1, dir1);
+
+        int rc = chdir(dir2.characters());
+        if (rc < 0) {
+            printf("chdir(%s) failed: %s", dir2.characters(), strerror(errno));
+            return 1;
+        }
+
+        g.cwd = dir2;
+
+        return 0;
+    }
+
+    // Let's assume the user's typed in 'pushd <dir>'
+    if (argc == 2) {
+        if (argv[1][0] == '/')
+            sprintf(pathbuf, "%s", argv[1]);
+        else
+            sprintf(pathbuf, "%s/%s", g.cwd.characters(), argv[1]);
+    } else if (argc == 3) {
+        for (int i = 0; i < argc; i++) {
+            const char* arg = argv[i];
+
+            if (arg[0] != '-') {
+                if (arg[0] == '/')
+                    sprintf(pathbuf, "%s", arg);
+                else
+                    sprintf(pathbuf, "%s/%s", g.cwd.characters(), arg);
+
+                dbg() << "hmmm " << arg;
+            }
+
+            if (!strcmp(arg, "-n"))
+                should_switch = false;
+        }
+    }
+
+    FileSystemPath canonical_path(pathbuf);
+    if (!canonical_path.is_valid()) {
+        printf("FileSystemPath failed to canonicalize '%s'\n", pathbuf);
+        return 1;
+    }
+
+    const char* real_path = canonical_path.string().characters();
+    g.directory_stack.append(real_path);
+
+    struct stat st;
+    int rc = stat(real_path, &st);
+    if (rc < 0) {
+        printf("stat(%s) failed: %s\n", real_path, strerror(errno));
+        return 1;
+    }
+
+    if (!S_ISDIR(st.st_mode)) {
+        printf("Not a directory: %s\n", real_path);
+        return 1;
+    }
+
+    if (should_switch) {
+        int rc = chdir(real_path);
+        if (rc < 0) {
+            printf("chdir(%s) failed: %s\n", real_path, strerror(errno));
+            return 1;
+        }
+
+        g.cwd = canonical_path.string();
+    }
+
+    return 0;
+}
+
+static int sh_dirs(int argc, char** argv)
+{
+    // The first directory in the stack is ALWAYS the current directory
+    g.directory_stack.at(0) = g.cwd.characters();
+
+    if (argc == 1) {
+        for (String dir : g.directory_stack)
```

Comment:
Prefer iterating with `auto&` to avoid creating a copy of everything iterated over:
```suggestion
        for (auto& directory : g.directory_stack)
```